Article: Manage software with GitOps (#28688)
fleetdm/fleet: articles/gitops-mode-software.md

# Manage software using GitOps

This guide is a walkthrough on how to manage software using GitOps. In Fleet, there are three types of software: [Fleet-maintained apps](#fleet-maintained-apps), [App Store (VPP) apps](#app-store-vpp-apps), and [custom packages](#custom-packages).

When using a git repository to manage Fleet, you can optionally put the UI in [GitOps mode](https://fleetdm.com/guides/gitops-mode). This prevents you from making changes in the UI that would be overridden by GitOps workflows.

## Fleet-maintained apps

TODO: Link out to reference docs for all options

## App Store (VPP) apps

TODO: Link to GitOps reference docs for all options

## Custom packages

> TODO: Instructions on how to use UI in GitOps mode

Reviewer: ```suggestion
1. In GitOps mode, select a team and navigate to **Software > Add Software > Custom package**
2. Choose a file to upload and select **Add software**
3. A modal will appear with YAML instructions.
    1. Create a YAML file with the suggested filename and populate it with the contents below.
    2. Save this file to your repository.
    3. Make sure that the package YAML is referenced from your team YAML.
    4. Download the additional queries and scripts that are linked in the modal and add them to your repository. Make sure to use the paths listed in the contents area above.
```

# Manage software using GitOps

This guide is a walkthrough on how to manage software using GitOps. In Fleet, there are three types of software: [Fleet-maintained apps](#fleet-maintained-apps), [App Store (VPP) apps](#app-store-vpp-apps), and [custom packages](#custom-packages).

When using a git repository to manage Fleet, you can optionally put the UI in [GitOps mode](https://fleetdm.com/guides/gitops-mode). This prevents you from making changes in the UI that would be overridden by GitOps workflows.

## Fleet-maintained apps

TODO: Link out to reference docs for all options

## App Store (VPP) apps

TODO: Link to GitOps reference docs for all options

## Custom packages

> 1. In GitOps mode, select a team and navigate to **Software > Add Software > Custom package**

Reviewer: ```suggestion
To use the Fleet UI to manage custom packages, first turn GitOps mode on in **Settings > Integration > Change management**
1. Navigate to **Software > Add Software > Custom package**
```

# Manage software using GitOps

This guide is a walkthrough on how to manage software using GitOps. In Fleet, there are three types of software: [Fleet-maintained apps](#fleet-maintained-apps), [App Store (VPP) apps](#app-store-vpp-apps), and [custom packages](#custom-packages).

> When using a git repository to manage Fleet, you can optionally put the UI in [GitOps mode](https://fleetdm.com/guides/gitops-mode). This prevents you from making changes in the UI that would be overridden by GitOps workflows.

Reviewer 2: ```suggestion
When using Gitops to manage Fleet, you can optionally put the UI in [GitOps mode](https://fleetdm.com/guides/gitops-mode). This prevents you from making changes in the UI that would be overridden by GitOps workflows.
```

## Fleet-maintained apps

> TODO: Link out to reference docs for all options

Reviewer: ```suggestion
Currently, managing [Fleet-maintained apps](https://fleetdm.com/guides/fleet-maintained-apps) is only supported using Fleet's UI or [API](https://fleetdm.com/docs/rest-api/rest-api) (YAML coming soon).
```

## App Store (VPP) apps

> TODO: Link to GitOps reference docs for all options

Reviewer: ```suggestion
To configure App Store (VPP) apps, see the [Software section](https://github.com/fleetdm/fleet/blob/main/docs/Configuration/yaml-files.md#software) of Fleet's best practice [GitOps documentation](https://github.com/fleetdm/fleet/blob/main/docs/Configuration/yaml-files.md#gitops).
```

> Currently, managing [Fleet-maintained apps](https://fleetdm.com/guides/fleet-maintained-apps) is only supported using Fleet's UI or [API](https://fleetdm.com/docs/rest-api/rest-api) (YAML coming soon).

Reviewer 2: ```suggestion
Currently, managing [Fleet-maintained apps](https://fleetdm.com/guides/fleet-maintained-apps) is only supported using Fleet's UI or [API](https://fleetdm.com/docs/rest-api/rest-api) (Gitops support coming soon).
```
I think we should also specify that Gitops won't delete/modify FMAs configured in the UI

> Currently, managing [Fleet-maintained apps](https://fleetdm.com/guides/fleet-maintained-apps) is only supported using Fleet's UI or [API](https://fleetdm.com/docs/rest-api/rest-api) (Gitops support coming soon).

Reviewer: ```suggestion
Currently, managing [Fleet-maintained apps](https://fleetdm.com/guides/fleet-maintained-apps) is only supported using Fleet's UI or [API](https://fleetdm.com/docs/rest-api/rest-api) (Gitops support coming soon. Currently, Gitops won't delete/modify Fleet-maintained apps configured in the UI).
```

> Currently, managing [Fleet-maintained apps](https://fleetdm.com/guides/fleet-maintained-apps) is only supported using Fleet's UI or [API](https://fleetdm.com/docs/rest-api/rest-api) (Gitops support coming soon. Currently, Gitops won't delete/modify Fleet-maintained apps configured in the UI).

Author: @eugkuo we shipped Fleet-maintained apps via GitOps right?

## Custom packages

To use the Fleet UI to manage custom packages, first turn GitOps mode on in **Settings > Integration > Change management**

> 1. Navigate to **Software > Add Software > Custom package**

Reviewer: ```suggestion
1. Navigate to **Software** and select a team. Then select **Add Software > Custom package**
```

> To configure App Store (VPP) apps, see the [Software section](https://github.com/fleetdm/fleet/blob/main/docs/Configuration/yaml-files.md#software) of Fleet's best practice [GitOps documentation](https://github.com/fleetdm/fleet/blob/main/docs/Configuration/yaml-files.md#gitops).

Reviewer: ```suggestion
To configure App Store (VPP) apps, via GitOps, please see the [Software section](https://github.com/fleetdm/fleet/blob/main/docs/Configuration/yaml-files.md#software) of Fleet's best practice [GitOps documentation](https://github.com/fleetdm/fleet/blob/main/docs/Configuration/yaml-files.md#gitops).
```

> 2. Choose a file to upload and select **Add software**

Reviewer: ```suggestion
2. Select a team and choose a file to upload and select **Add software**
```

> 2. Select a team and choose a file to upload and select **Add software**

Reviewer: ```suggestion
2. Choose a file to upload and select **Add software**
```

> To configure App Store (VPP) apps, via GitOps, please see the [Software section](https://github.com/fleetdm/fleet/blob/main/docs/Configuration/yaml-files.md#software) of Fleet's best practice [GitOps documentation](https://github.com/fleetdm/fleet/blob/main/docs/Configuration/yaml-files.md#gitops).

Reviewer 2: Also note that they have to add the vpp app to Apple Business Manager first

> To configure App Store (VPP) apps, via GitOps, please see the [Software section](https://github.com/fleetdm/fleet/blob/main/docs/Configuration/yaml-files.md#software) of Fleet's best practice [GitOps documentation](https://github.com/fleetdm/fleet/blob/main/docs/Configuration/yaml-files.md#gitops).

Reviewer: ```suggestion
To configure App Store (VPP) apps, via GitOps, please see the [Software section](https://github.com/fleetdm/fleet/blob/main/docs/Configuration/yaml-files.md#software) of Fleet's best practice [GitOps documentation](https://github.com/fleetdm/fleet/blob/main/docs/Configuration/yaml-files.md#gitops). Note that VPP apps must first be added to Apple Business Manager.
```

> To use the Fleet UI to manage custom packages, first turn GitOps mode on in **Settings > Integration > Change management**

Reviewer 2: What if we added why you would want to upload in the UI (there is no accessible URL)?

Reviewer: @mostlikelee I feel like if you're reading this and want to do it you'd know why you'd want to do it? Also, there might be other reasons someone would want to do this that we don't know and suggesting why might limit the imagination?

> To configure App Store (VPP) apps, via GitOps, please see the [Software section](https://github.com/fleetdm/fleet/blob/main/docs/Configuration/yaml-files.md#software) of Fleet's best practice [GitOps documentation](https://github.com/fleetdm/fleet/blob/main/docs/Configuration/yaml-files.md#gitops). Note that VPP apps must first be added to Apple Business Manager.

Reviewer: @mostlikelee added bit about ABM.

> To configure App Store (VPP) apps, via GitOps, please see the [Software section](https://github.com/fleetdm/fleet/blob/main/docs/Configuration/yaml-files.md#software) of Fleet's best practice [GitOps documentation](https://github.com/fleetdm/fleet/blob/main/docs/Configuration/yaml-files.md#gitops). Note that VPP apps must first be added to Apple Business Manager.

Author: ```suggestion
To configure App Store (VPP) apps, via GitOps, please see the [`app_store_apps`](https://github.com/fleetdm/fleet/blob/main/docs/Configuration/yaml-files.md#app_store_apps) of Fleet's best practice [GitOps documentation](https://github.com/fleetdm/fleet/blob/main/docs/Configuration/yaml-files.md#gitops). Note that VPP apps must first be added to [Apple Business Manager](https://business.apple.com).
```

I think we can link directly to the App Store apps section: https://fleetdm.com/docs/configuration/yaml-files#app-store-apps

Also a good opportunity to link directly to the Apple Business Manager login page.

> To use the Fleet UI to manage custom packages, first turn GitOps mode on in **Settings > Integration > Change management**

Author: ```suggestion
To configure custom packages via GitOps, please see the [`packages`](https://fleetdm.com/docs/configuration/yaml-files#packages) of Fleet's best practice [GitOps documentation](https://github.com/fleetdm/fleet/blob/main/docs/Configuration/yaml-files.md#gitops).

If you want to use Fleet to host custom packages instead of a third-party package hosting tool (ex. [Artifactory](https://jfrog.com/artifactory/)), first turn GitOps mode on in **Settings > Integration > Change management**
```

I agree with Tim that we want to include the why here. Maybe something like the above.

3. A modal will appear with YAML instructions.
    1. Create a YAML file with the suggested filename and populate it with the contents below.
    2. Save this file to your repository.
    3. Make sure that the package YAML is referenced from your team YAML.
    4. Download the additional queries and scripts that are linked in the modal and add them to your repository. Make sure to use the paths listed in the contents area above.

> TOOD: Link to GitOps reference docs for all options

Reviewer: ```suggestion
To learn more about configuring software using GitOps, [click here](https://github.com/fleetdm/fleet/blob/main/docs/Configuration/yaml-files.md#software).
```

> To learn more about configuring software using GitOps, [click here](https://github.com/fleetdm/fleet/blob/main/docs/Configuration/yaml-files.md#software).

Reviewer: ```suggestion
To learn more about configuring custom packages using GitOps, [click here](https://github.com/fleetdm/fleet/blob/main/docs/Configuration/yaml-files.md#software).
```

# Manage software using GitOps

This guide is a walkthrough on how to manage software using [GitOps](https://fleetdm.com/docs/configuration/yaml-files#gitops). In Fleet, there are three types of software: [Fleet-maintained apps](#fleet-maintained-apps), [App Store (VPP) apps](#app-store-vpp-apps), and [custom packages](#custom-packages).

When using Gitops to manage Fleet, you can optionally put the UI in [GitOps mode](https://fleetdm.com/guides/gitops-mode). This prevents you from making changes in the UI that would be overridden by GitOps workflows.

## Fleet-maintained apps

> Currently, managing [Fleet-maintained apps](https://fleetdm.com/guides/fleet-maintained-apps) is only supported using Fleet's UI or [API](https://fleetdm.com/docs/rest-api/rest-api) (Gitops support coming soon. Currently, Gitops won't delete/modify Fleet-maintained apps configured in the UI).

Author: ```suggestion
Currently, managing [Fleet-maintained apps](https://fleetdm.com/guides/fleet-maintained-apps) is only supported using Fleet's UI or [API](https://fleetdm.com/docs/rest-api/rest-api).
```

@eugkuo we changed this right? As of the latest Fleet, we will delete Fleet-maintained apps added in the UI.
